### PR TITLE
Simplify PandasDataset.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

File: src/gluonts/dataset/pandas.py [hunk 1]
```diff
@@ -57,8 +57,8 @@ class PandasDataset:
     static_features
         ``pd.DataFrame`` containing static features for the series. The index
         should contain the key of the series in the ``dataframes`` argument.
-    ignore_last_n_targets
-        For target and past dynamic features last ``ignore_last_n_targets``
+    future_length
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py:63]: 👍

Comment (lostella) [reply src/gluonts/dataset/pandas.py:63]: Let's mark the PR as breaking then

File: src/gluonts/dataset/pandas.py [hunk 2]
```diff
@@ -84,67 +84,66 @@ class PandasDataset:
...
 
-        assert isinstance(self._pairs, SizedIterable)
+        assert isinstance(pairs, SizedIterable)
+        self._data_entries = Map(self._pair_to_dataentry, pairs)
+
+        if self.feat_dynamic_real is None:
+            self.feat_dynamic_real = pd.DataFrame()
+
+        if self.past_feat_dynamic_real is None:
+            self.past_feat_dynamic_real = pd.DataFrame()
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py]: Does the typing of these attributes need to be adjusted?

Comment (jaheba) [reply src/gluonts/dataset/pandas.py]: The problem is that one can pass a `None`, but internally it is guaranteed to be not None.

Maybe it's easier to not make this a dataclass.

Comment (lostella) [reply src/gluonts/dataset/pandas.py]: But then I don't understand the change: this is supposed to be a list of column names, here it gets turned into a dataframe if it was None, what happens in the other case?

File: src/gluonts/dataset/pandas.py [hunk 3]
```diff
@@ -84,71 +84,63 @@ class PandasDataset:
...
     def num_feat_dynamic_real(self):
-        if self.feat_dynamic_real is None:
-            return 0
         return len(self.feat_dynamic_real)
 
     @property
     def num_past_feat_dynamic_real(self):
-        if self.past_feat_dynamic_real is None:
-            return 0
         return len(self.past_feat_dynamic_real)
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py]: These changes work as well if `self.feat_dynamic_real` is forced to be `[]` when found `None` (same for the other)

File: src/gluonts/dataset/pandas.py [hunk 4]
```diff
@@ -84,71 +84,63 @@ class PandasDataset:
...
 
-    @property
-    def cardinalities(self):
-        return [
-            max(self._static_cats[c]) + 1 for c in self._static_cats.columns
-        ]
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py]: Why remove this? I think this is valuable information

File: src/gluonts/dataset/pandas.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/dataset/pandas.py]: I think `dataclass` makes sense in this case, I'm not yet ready to give that up

File: src/gluonts/dataset/pandas.py [hunk 5]
```diff
@@ -50,15 +53,15 @@ class PandasDataset:
         Frequency of observations in the time series. Must be a valid pandas
         frequency.
     feat_dynamic_real
-        List of column names that contain dynamic real features.
+        list of column names that contain dynamic real features.
     past_feat_dynamic_real
-        List of column names that contain dynamic real features only available
+        list of column names that contain dynamic real features only available
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py]: not sure I agree, it's a word (not a code expression) starting a sentence

Comment (jaheba) [reply src/gluonts/dataset/pandas.py]: I think I accidentally changed this as well.

File: src/gluonts/dataset/pandas.py [hunk 6]
```diff
@@ -171,31 +169,30 @@ def _pair_to_dataentry(self, pair) -> DataEntry:
...
+        target = df[self.target].values
+        target = target[: len(target) - self.future_length]
+        entry["target"] = target.T
+
         if item_id is not None:
             entry["item_id"] = item_id
 
-        if self.num_feat_static_cat > 0:
-            entry["feat_static_cat"] = self._static_cats.loc[item_id].values
+        if self.num_feat_static_cat:
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py]: I prefer `self.num_feat_static_cat > 0` (not a fan of implicit conversions to `bool`)

Comment (jaheba) [reply src/gluonts/dataset/pandas.py]: But that's the pythonic way to do it.

File: src/gluonts/dataset/pandas.py [hunk 7]
```diff
@@ -207,23 +204,24 @@ def __len__(self) -> int:
...
     def __str__(self) -> str:
-        return (
-            f"PandasDataset<"
-            f"size={len(self)}, "
-            f"freq={self.freq}, "
-            f"num_dynamic_real={self.num_feat_dynamic_real}, "
-            f"num_past_dynamic_real={self.num_past_feat_dynamic_real}, "
-            f"num_static_real={self.num_feat_static_real}, "
-            f"num_static_cat={self.num_feat_static_cat}, "
-            f"cardinalities={self.cardinalities}>"
```
Comment (lostella) [inline src/gluonts/dataset/pandas.py:218]: Why not printing the cardinalities? I think it can be useful for quick validation

**Review by lostella — APPROVED** [review]

**Pull request merged**